dnsproxy: Use original source address and port

jrajahalme opened the pull request:
Make Cilium DNS proxy transparent by using the original source address (and port) in upstream connections. Since clients can issue multiple requests from the same port without waiting for the responses in between, we must support multiple requests in flight at the same time on the client side as well. To this end we share the DNS client between all requests that use the same upstream 5-tuple.

This new behaviour is controlled via a new  '--dnsproxy-enable-transparent-mode' option that defaults to 'false' for full backwards compatibility, but in Helm it will be defaulted to 'true' in non-chaining modes for new installs and upgrades from Cilium 1.12 onwards. Defaults can be overridden with explicit setting of 'dnsProxy.enableTransparentMode' to 'true' or 'false'. CNI chaining datapath mode is currently not compatible with the original source address use in Cilium proxies.

Support for shared clients is imported from a new version of cilium/dns. This version updates cilium/dns to avoid busy looping on receive errors.

Note for reviewers: 2nd commit was added to opt out from using the original source address when known to cause bind errors and/or failed connections due to 5-tuple collisions within the host networking namespace, or when masquerading is needed, but possibly skipped:
- when the source is in the host networking namespace (this was already done)
  - source is localhost
  - source is a host endpoint
- when the destination is in the host networking namespace
  - destination address is the node address
- when destination is outside of the cluster
  - destination security identity is the `WORLD`
  - destination security identity is a locally assigned CIDR identity

```release-note
Cilium DNS proxy can now use the original pod's address as the source address towards the DNS servers (--dnsproxy-enable-transparent-mode).
```

jrajahalme committed:
dnsproxy: Do not use original source when not possible

Do not use original source for server running in the local node, or when
the destination is outside of the cluster, as there is a risk of missing
masquarade on the upstream connection.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
dnsproxy: Use original source address in connections to dns servers

Set transparent, reuseaddr, and reuseport options and use the original
source address on connections from DNS proxy to DNS servers to allow use
of non-local source address as well as recreate sockets on the same
5-tuple without needing to wait for the TCP TIME_WAIT to finish.

Use the MagicMarkEgress mark on connections to the dns servers instead
the generic MagicMarkIdentity.

Use original source address in connections to dns servers when the source
address is not one of the host IPs. The original source address and port
can not be reused if there is already socket with them to the same
destination on the same networking namespace.

Use new dns.SharedClients to reuse DNS clients between all requests that
originate from the same source address and port. This allows multiple
different requests to be pending at the same time on the same dns Client,
which happens whenever the source pod sends multiple DNS requests from
the same resolver invocation, e.g., for A and AAAA records.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

aanm commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -916,33 +995,43 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
+	if !ep.IsHost() && !epAddr.IsLoopback() && !targetServerIsWorld && !targetServerID.HasLocalScope() && !targetServerIsLocal {
+		dialer.LocalAddr = originalSrcAddr
+		key = protocol + "-" + originalSrcAddr.String() + "-" + targetServerAddrStr
+	}
```
I think it would be important to leave a comment on these lines. They are kind of the more relevant lines in the entire PR.

jrajahalme replied:
Added back the comment, even though the comment was (and still is) inverted in its logic vs. the code below. But it may be helpful to think about this from both angles!

gandro commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -916,33 +995,43 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
-		p.sendRefused(scopedLog, w, request)
-		return
+	var key string
+	if !ep.IsHost() && !epAddr.IsLoopback() && !targetServerIsWorld && !targetServerID.HasLocalScope() && !targetServerIsLocal {
```
The `targetServerID.HasLocalScope()` seems to test if the target is outside of the cluster by checking if it is a CIDR identity. However, these days, CIDR identites can [already overlap with node identities](https://github.com/cilium/cilium/pull/27464) and soon will probably also overlap with pod identities. So `targetServerID.HasLocalScope()` might be true for remote pods in the future. I think we ought to introduce a proper helper here that is a bit more future proof.

Do we have CI coverage for those cases (i.e. testing DNS traffic to a pod, to a CIDR identity, to a remote node)? Otherwise I'm afraid we might introduce regressions, because "local identity" might no longer mean "outside of cluster".

jrajahalme replied:
We likely do have CI coverage gaps here. Added a new helper `identity.IsCluster()` to keep this semantic in the identity package.

gandro commented on pkg/ipcache/ipcache.go:
```diff
@@ -806,18 +807,46 @@ func (ipc *IPCache) LookupByPrefix(IP string) (Identity, bool) {
+			local = ip4.Equal(node.GetIPv4())
+		} else if hostIP != nil {
+			local = hostIP.Equal(node.GetIPv6())
+		}
```
So the "hostIP" here is the tunnel IP, right? We don't always set the tunnel IP, depending on IP type, encryption, hostfirewall etc:

https://github.com/cilium/cilium/blob/076653691d0d15b4449d6e2791bdbaea5298ed5f/pkg/node/manager/manager.go#L389-L396

I don't think relying on the tunnel IP is robust. We should instead rely on the IP identity (i.e. is it localhost) and on the endpoint list.

jrajahalme replied:
You are right, and as you suggested elsewhere, checking against the reserved host identity suffices here and is much simpler!

gandro reviewed (changes requested):
Didn't review the vendored in changes yet, but I have some questions about IPCache usage

gandro reviewed (approved):
Rebase on main is needed to fix the unit test failures, but other than this seems to look good to me now

rolinh reviewed (approved):
Vendor changes lgtm

pippolo84 reviewed (approved):
LGTM, thanks! :100:

jrajahalme committed:
option: Add --dnsproxy-enable-transparent-mode (default false)

Add dnsproxy-enable-transparent-mode option to enable DNS Proxy
transparent mode. If 'true', Cilium DNS proxy will use the original
source address of the source pod in the forwarded DNS requests. Local
host sources and destinations are excepted due to networking stack
compatibility reasons, but the use of the original address is typically
not significant for node local traffic.

Defaults to 'false' for backwards compatibility for upgrades, or to
'true' for Cilium 1.12 onwards.

Transparent mode is not compatible with CNI chaning modes, so if CNI
chaining is used, transparent mode will not be set unless explicitly set
with helm value 'dnsProxy.enableTransparentMode=true'.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

danehans commented on daemon/cmd/daemon_main.go:
```diff
@@ -885,6 +885,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.MarkHidden(option.DNSProxyLockTimeout)
 	option.BindEnv(vp, option.DNSProxyLockTimeout)
 
+	flags.Bool(option.DNSProxyEnableTransparentMode, defaults.DNSProxyEnableTransparentMode, "Enable DNS proxy transparent mode")
```
Are any other modes possible in the future? If so, it would be better to use an enum string, e.g. `legacy`, `transparent`, etc.

jrajahalme replied:
`cilium-envoy` has been running in transparent mode for years now, I don't see that changing anytime soon. This PR changes our DNS proxy to the same behavior, so a simple boolean is OK here.

danehans commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -48,6 +49,7 @@
       {{- $azureUsePrimaryAddress = "false" -}}
   {{- end }}
   {{- $defaultKubeProxyReplacement = "disabled" -}}
+  {{- $defaultDNSProxyEnableTransparentMode = "true" -}}
```
Why is the default different for v1.12 or later?

jrajahalme replied:
v1.12 is the oldest supported version, and this change is planned to be backported to v1.12, v1.13, v1.14, hence the default value `true` for these versions. There is no plan to change the behavior if upgrade compatibility is specified as v1.11 or older.

gandro replied:
~I think this is the wrong way around. If we want to not enable the feature upon upgrade, we ought to set it to false here.~

Edit: The code is correct, since all `upgradeCompatibility` blocks are executed even if the value is unset.

danehans commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -762,13 +766,88 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
+//     use the original port number.
+//
+//   - (DNS) clients use ephemeral source ports, i.e., the port can be different in every
+//     request. Typically, a DNS resolver library uses the same ephemeral port only to requests from
```
Should `to` be `for`?

jrajahalme replied:
Yes, fixed!

danehans commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -762,13 +766,88 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
+//     for years already) when a client pod has rapidly cycled through its ephemeral port space,
+//     e.g. when performing netperf CRR or similar performance tests.
+//
+//   - We could try keep the client and its bound socket around for some minimal time to save
```
s/`keep`/`keeping`/

jrajahalme replied:
Grammar check told to write this as `to keep`, done!

danehans reviewed (commented):
Should `proxy_test.go` be updated to include unit tests for this PR?

nathanjsweet reviewed (approved):
LGTM.

gandro reviewed (changes requested):
NACK for Helm, the `upgradeCompatibility` logic seems reversed

gandro reviewed (approved):
Discussed offline, I confused myself about `upgradeCompatibility` once again. The Helm logic as is works as advertised and seems sound.

Goal is to always enable it (which the `upgradeCompatibility` section does, because those are also taken if the value is unset), unless in chaining mode or if the user explicitly disabled it. The current code achieves that.
